libsimulator/src/AnticipationVelocityModel.cpp

```diff
@@ -0,0 +1,369 @@
+    constexpr double anticipationTimeMin = 0.0;
+    constexpr double anticipationTimeMax = 5.0;
+    validateConstraint(
+        anticipationTime, anticipationTimeMin, anticipationTimeMax, "anticipationTime");
+
+    const auto reactionTime = model.reactionTime;
+    constexpr double reactionTimeMin = 0.0;
+    constexpr double reactionTimeMax = 1.0;
+    validateConstraint(reactionTime, reactionTimeMin, reactionTimeMax, "reactionTime", true);
+
+    const auto neighbors = neighborhoodSearch.GetNeighboringAgents(agent.pos, 2);
+    for(const auto& neighbor : neighbors) {
+        if(agent.id == neighbor.id) {
+            continue;
+        }
+        const auto& neighbor_model = std::get<AnticipationVelocityModelData>(neighbor.model);
+        const auto contanctdDist = r + neighbor_model.radius;
+        const auto distance = (agent.pos - neighbor.pos).Norm();
+        if(contanctdDist >= distance) {
+            throw SimulationError(
+                "Model constraint violation: Agent {} too close to agent {}: distance {}",
+                agent.pos,
+                neighbor.pos,
+                distance);
+        }
+    }
+
+    const auto lineSegments = geometry.LineSegmentsInDistanceTo(r, agent.pos);
+    if(std::begin(lineSegments) != std::end(lineSegments)) {
+        throw SimulationError(
+            "Model constraint violation: Agent {} too close to geometry boundaries, distance "
+            "<= {}",
+            agent.pos,
+            r);
+    }
+}
+
+std::unique_ptr<OperationalModel> AnticipationVelocityModel::Clone() const
+{
+    return std::make_unique<AnticipationVelocityModel>(*this);
+}
+
+double AnticipationVelocityModel::OptimalSpeed(
+    const GenericAgent& ped,
+    double spacing,
+    double time_gap) const
+{
+    const auto& model = std::get<AnticipationVelocityModelData>(ped.model);
+    double min_spacing = 0.0;
+    return std::min(std::max(spacing / time_gap, min_spacing), model.v0);
+}
+
+double AnticipationVelocityModel::GetSpacing(
+    const GenericAgent& ped1,
+    const GenericAgent& ped2,
+    const Point& direction) const
+{
+    const auto& model1 = std::get<AnticipationVelocityModelData>(ped1.model);
+    const auto& model2 = std::get<AnticipationVelocityModelData>(ped2.model);
+    const auto distp12 = ped2.pos - ped1.pos;
+    const auto inFront = direction.ScalarProduct(distp12) >= 0;
+    if(!inFront) {
+        return std::numeric_limits<double>::max();
+    }
+
+    const auto left = direction.Rotate90Deg();
+    const auto l = model1.radius + model2.radius;
+    bool inCorridor = std::abs(left.ScalarProduct(distp12)) <= l;
+    if(!inCorridor) {
+        return std::numeric_limits<double>::max();
+    }
+    return distp12.Norm() - l;
+}
+
+Point AnticipationVelocityModel::CalculateInfluenceDirection(
+    const Point& desiredDirection,
+    const Point& predictedDirection) const
+{
+    // Eq. (5)
+    static std::mt19937 gen(42);
```

Comment:
The seed needs to become a model parameter. Otherwise I think this will become a breaking change later on.